crates/transaction-emitter-lib/src/cluster.rs

```diff
@@ -218,3 +231,91 @@ impl Cluster {
         self.instances.iter()
     }
 }
+
+fn get_most_frequent_chain_id_from_instances(instance_states: Vec<(Instance, State)>) -> ChainId {
+    // If the chain_id is not provided, we use the most common chain_id from the instances
+    // assuming that if there is a mistake in providing the targets, then there is still a majority
+    // that has the right chain_id.
+    let chain_id_u8 = instance_states
```

Comment:
Hmm this is doing an O(n2) operation across instance states. Can we do it in O(n)?

Reply:
sg, updated. We haven't run txn-emitter against too many instances though, but good to optimize ;)